topydo/ui/columns/Transaction.py

```diff
@@ -0,0 +1,59 @@
+# Topydo - A todo.txt client written in Python.
+# Copyright (C) 2015 Bram Schoenmakers <bram@topydo.org>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from topydo.lib.MultiCommand import MultiCommand
+
+
+class Transaction(object):
+    """
+    This class implements basic handling of executing any subcommand on multiple
+    todo items.
+    """
+    def __init__(self, p_subcommand=None, p_todo_ids=None):
+        self._subcommand = p_subcommand
+        self._todo_ids = p_todo_ids
+        self._operations = []
+
+    def prepare(self, p_args):
+        """
+        Prepares list of operations to execute based on p_args, list of
+        todo items contained in _todo_ids attribute and _subcommand
+        attribute.
+        """
+        if self._todo_ids:
+            id_position = p_args.index('{}')
+            real_subcommand = self._subcommand([])
+
+            if isinstance(real_subcommand, MultiCommand):
+                p_args[id_position:id_position + 1] = self._todo_ids
+                self._operations.append(p_args)
+            else:
+                for todo_id in self._todo_ids:
+                    operation_args = p_args[:]
+                    operation_args[id_position] = todo_id
+                    self._operations.append(operation_args)
+        else:
+            self._operations.append(p_args)
+
+    def execute(self):
+        """
+        Executes each operation from _operations attribute.
+        """
+        last_operation = len(self._operations) - 1
+        for i, operation in enumerate(self._operations):
+            command = self._subcommand(operation)
+            if command.execute() != False and i == last_operation:
```

Comment:
> What if one of the intermediate operations fails (return value False)? Also, when the last operation fails, execute() will have no return value (should return False).

That's a valid concern.

> Moreover, in terminology, a transaction is something that either fails and rolls back or completes successfully. This transaction is a best effort: try to succeed as many items as we can.

That shouldn't be so hard to implement since in case of any fail we can in my opinion just call `self.backup.apply(self.todolist)` from `UIApplication`.

> I do wonder how likely it is that a command fails for some items and succeeds for others. The command is the same everytime, except for the todo ID. And the todo ID is not chosen by the user but passed on by topydo.

In my opinion it is practically imposible with one topydo instance and without any other todo app connected to the same todo.txt file. With concurrent topydo instances or any other app using same todo file – definetely possible:

1. text ids are on
2. mark at least two items
3. go to the command-line
4. in the other topydo instance delete/change (so the id will disappear or change) one of the marked items
5. go back to first instance where you marked items - you should be still in command-line, and one of the marked items is not there at all or is at least not marked anymore. Type: `append {} Foo`
6. There should be one line with updated todo item and `Invalid todo number given` - as You said: best effort.

It could be fixable by detecting if ids in `self.marked_todos` are accessible after update caused by watchdog or maybe (less subtly) just clear marks and kick the user out of the command-line after any watchdog caused update. Still, I would suggest to implement some kind of `self._rollback()`. Just in case.